### PR TITLE
fix: remove calls to Pytorch Dataset len

### PR DESCRIPTION
**Pull request by wes-turner**
## Description

Pytorch Datasets (`torch.utils.data.Dataset`) aren't guaranteed to have a `__len__` implemented (Datasets can be either "map-style" or "iterable-style". When map-style, they must implement a `__len__`, and when iterable-style they _may_). The `__len__` on a Pytorch DataLoader may pass the call through to its Dataset.

A `det.pytorch.PyTorchTrial` is typically constructed from a `det.pytorch.DataLoader`. `det.pytorch.DataLoader` cannot, itself, front an iterable-style Pytorch Dataset. It is, however, possible to construct a `det.pytorch.PyTorchTrial` with an unwrapped `torch.utils.data.Dataset` if `context.experimental.disable_dataset_reproducibility_checks()` is called in the `PyTorchTrial`'s `__init__`.

Before this patch, during a `PyTorchTrialContext.run` we called `len` on the trial's validation dataloader. Per the above, it had been possible to construct a trial with a validation dataloader that did not have `__len__` implemented, and in this case `run` would raise a runtime `TypeError` exception.

Turns out, though, those existing calls to `__len__` that weren't actually necessary. This patch revises them with no functional change in behavior

* instead of `len(validation_loader)` to check for emptiness before iterating through it, instead check the number of times the validation_loader is iterated through, raising the same error if it was empty.
* removes a call to `len` where the result was entirely ignored.

This PR also makes a couple "continuous improvement" changes, including moving around a couple pieces of code and renames variables so that its logic is a little more obvious.

## Test Plan

I've tested this by hand by modifying `build_validation_training_loader` in the example https://github.com/determined-ai/determined/blob/main/examples/tutorials/mnist_pytorch/train.py to return a Dataset for a `torch.utils.data.Dataset` subclass that has no implemented `__len__` and then running https://github.com/determined-ai/determined/blob/main/harness/tests/experiment/pytorch/test_examples.py tests on it. Without the patch, validation fails because of the call to `__len__`. With the patch, validation succeeds.

We don't have any unit tests of the function I modified (`_PyTorchTrialController._run` or its caller `_PyTorchTrialController.run`), and this doesn't quite seem like enough of a patch to create unit tests for the class for it. It also doesn't seem quite appropriate to create another end-to-end test just to ensure `__len__` isn't called on a validation loader. Maybe more automated tests aren't needed?

For the release party, if you'd like to test this yourself, create a `PyTorchTrial` object of a class that's implemented `build_validation_data_loader` to return a plain, unwrapped `torch.utils.data.DataLoader` that itself has no `__len__`. Then run a training loop for this trial object.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket

[MLG-1022]

### CODE REVIEW

File: harness/determined/pytorch/_pytorch_trial.py [hunk 1]
```diff
@@ -964,11 +962,10 @@ def _validate(self, searcher_op: Optional[core.SearcherOperation] = None) -> Dic
             batch_metrics = []
 
             assert isinstance(self.validation_loader, torch.utils.data.DataLoader)
-            if len(self.validation_loader) == 0:
-                raise RuntimeError("validation_loader is empty.")
             for callback in self.callbacks.values():
                 callback.on_validation_epoch_start()
 
+            idx = -1
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/_pytorch_trial.py:968]: You are taking a check which was fully contained in two consecutive lines and spreading it over a wider area here.  I don't want you to rewrite it because what you have is simple and easy, but can you just put a comment above `idx = -1` to explain why -1 is significant?

File: harness/determined/pytorch/_pytorch_trial.py [hunk 2]
```diff
@@ -1059,12 +1054,17 @@ def _validate(self, searcher_op: Optional[core.SearcherOperation] = None) -> Dic
             # common than evaluate_batch() and we can't know how the user processed their
             # validation data.
             if self._evaluate_batch_defined():
+                # Reshape and sum.
+                # TODO: remove the type directive once we upgrade to mypy >= 1.7.0
+                inputs_total, batches_total = [sum(n) for n in zip(*input_counts)]  # type: ignore
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/_pytorch_trial.py:1059]: input_counts doesn't seem to be defined in the evaluate_full_dataset codepath

File: harness/determined/pytorch/_pytorch_trial.py [hunk 3]
```diff
@@ -1553,20 +1553,30 @@ def train_batch(
...
+    ) -> Union[pytorch.DataLoader, torch.utils.data.DataLoader]:
         """
         Defines the data loader to use during validation.
 
-        Must return an instance of :py:class:`determined.pytorch.DataLoader`.
+        For full "determined" functionality, this must return an instance of
+        :py:class:`determined.pytorch.DataLoader`. It can also return an unwrapped
+        :py:class:`torch.utils.data.DataLoader` if you need more control over the underlying
+        DataLoader and are willing to sacrifice some Determined features (ex: automatic data
+        sharding).
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/_pytorch_trial.py:1579]: Thank you for catching that our typing and docstring had fallen out of correctness.

However, this is a user-facing docstring, and given that I think this addition is much too vague.  What is `full "determined" functionality` anyway?

I wouldn't recommend answering that question here; I would link to [our existing docs on the subject](https://docs.determined.ai/latest/model-dev-guide/api-guides/apis-howto/api-pytorch-ug.html#customize-a-reproducible-dataset).

Something like:

```
Users with a MapDataset will normally return a :class:`determined.pytorch.DataLoader`, but users with an IterableDataset or with other advanced needs may return a bare ``torch.utils.data.DataLoader`` if they follow the steps described in :ref:`pytorch-reproducible-dataset`.
```

File: harness/determined/pytorch/_pytorch_trial.py [hunk 4]
```diff
@@ -1553,20 +1553,30 @@ def train_batch(
...
-    def build_validation_data_loader(self) -> pytorch.DataLoader:
+    def build_validation_data_loader(
+        self,
+    ) -> Union[pytorch.DataLoader, torch.utils.data.DataLoader]:
         """
         Defines the data loader to use during validation.
 
-        Must return an instance of :py:class:`determined.pytorch.DataLoader`.
+        For full "determined" functionality, this must return an instance of
+        :py:class:`determined.pytorch.DataLoader`. It can also return an unwrapped
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/_pytorch_trial.py:1576]: skip the `:py` in `:py:class:...`, as it is not necessary

**Review by rb-determined-ai — COMMENTED** [review]
You are missing the most important usage of `len(data_loader)` which is [here](https://github.com/determined-ai/determined/blob/e2cf9809d66c492cb82e8a1311c7fa540a286094/harness/determined/pytorch/_pytorch_trial.py#L465-L472).

File: harness/determined/pytorch/_pytorch_trial.py (hunk 1 above)
Comment (wes-turner) [reply harness/determined/pytorch/_pytorch_trial.py:968]: That's a nice idea. Done.

File: harness/determined/pytorch/_pytorch_trial.py (hunk 2 above)
Comment (wes-turner) [reply harness/determined/pytorch/_pytorch_trial.py:1059]: The code is kind of unfortunately structured and pretty hard to figure out:
```
1. if evaluate_batch:
2.   do stuff in a managed way
3.   if chief:
4.     do managed stuff relevant to the chief
5. else:  # evaluate_full
6.   do whatever stuff evaluate_full says to do
7. 
8. if chief:
9.   if evaluate_batch:
10.     report batch-specific detail
11.  report general stuff
```    
Before patch:
* `num_inputs` defined / calculated line 2
* `input_counts` defined line 2 (from gathered `num_inputs`)
* `num_inputs` re-defined line 4 when evaluate_batch (from `input_counts`)
* `num_inputs` re-defined line 6 when evaluate_full (from `len(validation_loader)`)
* `num_inputs` (per its second definition) used line 10 and nowhere else

Pre-patch, `num_inputs` was being defined when both evaluate_batch and evaluate_full, but only used during evaluate_batch.

After patch:
* `num_inputs` defined / calculated line 2
* `input_counts` defined line 2 (from gathered `num_inputs`)
* `inputs_total` defined line 10 (from `input_counts`)
* `inputs_total` used line 10

Effects of this patch:
* `num_inputs` not defined during evaluate_full (which is fine, because it hadn't been used there)
* gathered calculation from `input_counts` moved to where it's used (line 10)
* gathered calculation from `input_counts` given a new name so `num_inputs` isn't overloaded

File: harness/determined/pytorch/_pytorch_trial.py (hunk 4 above)
Comment (wes-turner) [reply harness/determined/pytorch/_pytorch_trial.py:1576]: Nice. Thanks!

File: harness/determined/pytorch/_pytorch_trial.py (hunk 3 above)
Comment (wes-turner) [reply harness/determined/pytorch/_pytorch_trial.py:1579]: I agree about the vagueness. It made me feel a little gross, too.

Ideally, the docstring should say _why_ a user would choose which class to return. It's somewhat easy to say "if you can't return a `det.pytorch.DataLoader`, return the other one. It's harder to come up with an explanation for "why managed" that's not vague.

I like your idea of referring to docs for that. That's something docs.determined.ai should do. The doc you suggested does have a nice note explaining how, I can't find anything on the site for why.

I think a voice chat is probably the best way to work something out from here. When you get to this and you've got the time to, could you please give me a call?

Comment (wes-turner) [reply harness/determined/pytorch/_pytorch_trial.py]: Ended up with a little from column A, a little from column B.

**Review by azhou-determined — APPROVED** [review]
nice
